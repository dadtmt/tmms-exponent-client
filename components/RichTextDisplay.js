import React, { PropTypes } from 'react'
import { Text, View } from 'react-native'
import redraft from 'redraft'

let inlineKey = 0

const renderers = {
  blocks: {
    unstyled: (children) => children.map(
      (child, index) => <Text style={styles.unstyled} key={index} >{child}</Text>
    ),
    'header-one': (children) => children.map(
      (child, index) => <Text style={styles.h1} key={index} >{child}</Text>
    ),
    'header-two': (children) => children.map(
      (child, index) => <Text style={styles.h2} key={index} >{child}</Text>
    ),
    'unordered-list-item': (children) => children.map(
      (child, index) => <View style={{flexDirection: 'row'}} key={index} >
        <Text>{'\u2022'}</Text>
        <Text style={{flex: 1, paddingLeft: 5}}>{child}</Text>
      </View>
    )
  },
  inline: {
    BOLD: (children) => <Text style={styles.bold} key={inlineKey++} >{children}</Text>,
    ITALIC: (children) => <Text style={styles.italic} key={inlineKey++} >{children}</Text>,
    UNDERLINE: (children) => <Text style={styles.underline} key={inlineKey++} >{children}</Text>
  },
}

const RichTextDisplay = ({ rawContent }) => <View>
  {redraft(rawContent, renderers)}
</View>

RichTextDisplay.propTypes = {
  rawContent: PropTypes.object.isRequired
}

const styles = {
  bold: {
    fontWeight: 'bold'
  },
  h1: {
    fontSize: 40
  },
  h2: {
    fontSize: 25
  },
  italic: {
    fontStyle: 'italic'
  },
  underline: {
    textDecorationLine: 'underline'
  },
  unstyled: {
    fontSize: 15
  }
}

export default RichTextDisplay
