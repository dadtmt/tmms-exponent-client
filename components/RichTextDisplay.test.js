import React from 'react'
import renderer from 'react-test-renderer'

import RichTextDisplay from './RichTextDisplay'

it('renders content', () => {
  const props = {
    rawContent: {
      blocks: [
        {
          data: {},
          depth: 0,
          entityRanges: [],
          inlineStyleRanges: [],
          key: '4bkmd',
          text: 'some rich text',
          type: 'unstyled'
        }
      ],
      entityMap: {}
    }
  }
  expect(
    renderer.create(<RichTextDisplay {...props} />).toJSON()
  ).toMatchSnapshot()
})

it('renders bulletlist', () => {
  const props = {
    rawContent: {
      'entityMap': {},
      'blocks': [
        {
          'key': '1madj',
          'text': 'this is a list:',
          'type': 'unstyled',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': 'bd8au',
          'text': 'first item',
          'type': 'unordered-list-item',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': '1ogut',
          'text': 'second item',
          'type': 'unordered-list-item',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        }
      ]
    }
  }
  expect(
    renderer.create(<RichTextDisplay {...props} />).toJSON()
  ).toMatchSnapshot()
})

it('renders rich text content', () => {
  const props = {
    rawContent: {
      'entityMap': {},
      'blocks': [
        {
          'key': 'cbm0g',
          'text': 'titre H1',
          'type': 'header-one',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': '5tivm',
          'text': 'titre H2',
          'type': 'header-two',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': '5n20d',
          'text': 'unstyled paragraph',
          'type': 'unstyled',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': 'da4nc',
          'text': '',
          'type': 'unstyled',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': 'b3dsv',
          'text': 'this is italic style and this is bold style and this is underline style and this is all styles',
          'type': 'unstyled',
          'depth': 0,
          'inlineStyleRanges': [
            {
              'offset': 8,
              'length': 13,
              'style': 'ITALIC'
            },
            {
              'offset': 84,
              'length': 10,
              'style': 'ITALIC'
            },
            {
              'offset': 84,
              'length': 10,
              'style': 'BOLD'
            },
            {
              'offset': 84,
              'length': 10,
              'style': 'UNDERLINE'
            }
          ],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': '97hah',
          'text': '',
          'type': 'unstyled',
          'depth': 0,
          'inlineStyleRanges': [],
          'entityRanges': [],
          'data': {}
        },
        {
          'key': '68ht0',
          'text': 'h1 with styles',
          'type': 'header-one',
          'depth': 0,
          'inlineStyleRanges': [
            {
              'offset': 0,
              'length': 2,
              'style': 'ITALIC'
            },
            {
              'offset': 3,
              'length': 4,
              'style': 'BOLD'
            },
            {
              'offset': 8,
              'length': 6,
              'style': 'UNDERLINE'
            }
          ],
          'entityRanges': [],
          'data': {}
        }
      ]
    }
  }
  expect(
    renderer.create(<RichTextDisplay {...props} />).toJSON()
  ).toMatchSnapshot()
})
