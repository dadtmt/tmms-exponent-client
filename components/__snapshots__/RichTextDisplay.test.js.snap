exports[`test renders bulletlist 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    }>
    this is a list:
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      •
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "flex": 1,
          "paddingLeft": 5,
        }
      }>
      first item
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      •
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "flex": 1,
          "paddingLeft": 5,
        }
      }>
      second item
    </Text>
  </View>
</View>
`;

exports[`test renders content 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    }>
    some rich text
  </Text>
</View>
`;

exports[`test renders rich text content 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 40,
      }
    }>
    titre H1
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 25,
      }
    }>
    titre H2
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    }>
    unstyled paragraph
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    } />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    }>
    this is 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontStyle": "italic",
        }
      }>
      italic style 
    </Text>
    and this is bold style and this is underline style and this is 
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontStyle": "italic",
        }
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontWeight": "bold",
          }
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "textDecorationLine": "underline",
            }
          }>
          all styles
        </Text>
      </Text>
    </Text>
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 15,
      }
    } />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 40,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontStyle": "italic",
        }
      }>
      h1
    </Text>
     
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontWeight": "bold",
        }
      }>
      with
    </Text>
     
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "textDecorationLine": "underline",
        }
      }>
      styles
    </Text>
  </Text>
</View>
`;
